python/mxnet/gluon/data/dataloader.py

```diff
@@ -275,9 +284,24 @@ def __iter__(self):
     def shutdown(self):
         """Shutdown internal workers by pushing terminate signals."""
         if not self._shutdown:
+            # send shutdown signal to the fetcher and join data queue first
+            # Remark:   loop_fetcher need to be joined prior to the workers.
+            #           otherwise, the the fetcher may fail at getting data
+            self._data_queue.put((None, None))
+            self._fetcher.join()
+            # send shutdown signal to all worker processes
             for _ in range(self._num_workers):
                 self._key_queue.put((None, None))
-            self._data_queue.put((None, None))
+            # wait 0.5 sec for workers to shut down if necessary
+            for w in self._workers:
+                if w.is_alive():
+                    import time
+                    time.sleep(0.5)   # allow 0.5 sec to join
```

Comment:
sleep in the main thread is not a good idea.

Reply:
I have two options:
1) simply remove this loop. Then, do not wait the workers to finish, and just kill them if they are alive. 
2) fork a new process to do the waiting and termination.

Reply:
It's safe to go with 1.

Reply:
pushed with option 1